dmapiclient/base.py

```diff
@@ -49,17 +49,58 @@ def iter_method(self, *args, **kwargs):
     return iter_method
 
 
+class classproperty(object):
+    def __init__(self, getter):
+        self.getter = getter
+
+    def __get__(self, instance, owner):
+        return self.getter(owner)
+
+
 class BaseAPIClient(object):
-    RETRIES = 5
-    RETRIES_BACKOFF_FACTOR = 0.3
+    _RETRIES = 5
+    _RETRIES_BACKOFF_FACTOR = 0.3
     #  Respose status codes to retry on.
-    RETRIES_FORCE_STATUS_CODES = (500, 502, 503, 504)
+    _RETRIES_FORCE_STATUS_CODES = (500, 502, 503, 504)
+
+    # the following are really intended to be read-only from outside the class, hence properties
+    @classproperty
+    def RETRIES(cls):
+        return cls._RETRIES
+
+    @classproperty
+    def RETRIES_BACKOFF_FACTOR(cls):
+        return cls._RETRIES_BACKOFF_FACTOR
+
+    @classproperty
+    def RETRIES_FORCE_STATUS_CODES(cls):
+        return cls._RETRIES_FORCE_STATUS_CODES
+
+    @property
+    def base_url(self):
+        return self._base_url
+
+    @property
+    def auth_token(self):
+        return self._auth_token
+
+    @property
+    def enabled(self):
+        return self._enabled
+
+    @property
+    def timeout(self):
+        return self._timeout
+
+
+    __slots__ = ("_base_url", "_auth_token", "_enabled", "_timeout",)
```

Comment:
This might be a dumb question: what is this used for?

Reply:
Giving a python class `__slots__` tells python not to give it a dictionary for implementing its attributes, instead giving it a fixed selection of attributes. Basically, "don't let people assign attributes on an instance other than these listed"

Reply:
Aha, thanks! Every day is a school day.

Reply:
hmm unfortunately it seems to make `mock.patch.object` very unhappy.